src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -403,23 +405,47 @@ public static void CheckAllConstraints(this TypeSymbol type, CheckConstraintsArg
             public readonly bool IncludeNullability;
             public readonly Location Location;
             public readonly BindingDiagnosticBag Diagnostics;
+            public readonly CompoundUseSiteInfo<AssemblySymbol> Template;
 
             public CheckConstraintsArgs(CSharpCompilation currentCompilation, ConversionsBase conversions, Location location, BindingDiagnosticBag diagnostics) :
                 this(currentCompilation, conversions, currentCompilation.IsFeatureEnabled(MessageID.IDS_FeatureNullableReferenceTypes), location, diagnostics)
             {
             }
 
-            public CheckConstraintsArgs(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics)
+            public CheckConstraintsArgs(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics) :
+                this(currentCompilation, conversions, includeNullability, location, diagnostics, template: new CompoundUseSiteInfo<AssemblySymbol>(diagnostics, currentCompilation.Assembly))
+            {
+            }
+
+            public CheckConstraintsArgs(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics, CompoundUseSiteInfo<AssemblySymbol> template)
             {
                 this.CurrentCompilation = currentCompilation;
                 this.Conversions = conversions;
                 this.IncludeNullability = includeNullability;
                 this.Location = location;
                 this.Diagnostics = diagnostics;
+                this.Template = template;
             }
         }
 
-        private static readonly Func<TypeSymbol, CheckConstraintsArgs, bool, bool> s_checkConstraintsSingleTypeFunc = (type, arg, unused) => CheckConstraintsSingleType(type, arg);
+        internal class CheckConstraintsArgsBoxed
```

Comment:
>CheckConstraintsArgsBoxed [](start = 23, length = 25)

nit: consider the same naming/organization we used for `TypeWithAnnotations.Boxed`
Also, it may be worth leaving a comment saying this type is mostly to save stack space (as opposed to atomic initialization as we did for `TypeWithAnnotations.Boxed`).

Reply:
> it may be worth leaving a comment saying this type is mostly to save stack space (as opposed to atomic initialization as we did for TypeWithAnnotations.Boxed).

I do not think comments like that are useful. It is OK to use types in any way that serves the purpose of the usage. It is a tool, and, if it does the job, it does the job. Examples are sometimes useful, but here we have them locally (in this file).

---
In reply to: [420412907](https://github.com/dotnet/roslyn/pull/43893#discussion_r420412907) [](ancestors = 420412907)